File: src/Blokemon.Game/AuthorityCatalog.cs
using Blokemon.Core.SetDesign;

namespace Blokemon.Game;

internal sealed class AuthorityCatalog
{
    private readonly Dictionary<string, BlokemonCollectible> _collectibles;
    private readonly Dictionary<string, BlokemonKit> _kits;
    private readonly Dictionary<string, BlokemonBasicVim> _vim;
    private readonly Dictionary<string, BlokemonAttack> _attacks;
    private readonly Dictionary<string, BlokemonPartyTrick> _partyTricks;
    private readonly Dictionary<string, BlokemonHouseRule> _houseRules;

    public AuthorityCatalog(BlokemonRuntimeManifest manifest)
    {
        Manifest = manifest;
        _collectibles = manifest.Collectibles.ToDictionary(static card => card.Id);
        _kits = manifest.Kits.ToDictionary(static card => card.Id);
        _vim = manifest.BasicVim.ToDictionary(static card => card.Id);
        _attacks = manifest
            .Collectibles.SelectMany(static card => card.Attacks)
            .Concat(manifest.Kits.SelectMany(static card => card.Attacks))
            .ToDictionary(static effect => effect.MechanicalId);
        _partyTricks = manifest
            .Collectibles.SelectMany(static card => card.PartyTricks)
            .Concat(manifest.Kits.SelectMany(static card => card.PartyTricks))
            .ToDictionary(static effect => effect.MechanicalId);
        _houseRules = manifest
            .Collectibles.SelectMany(static card => card.HouseRules)
            .Concat(manifest.Kits.SelectMany(static card => card.HouseRules))
            .ToDictionary(static effect => effect.MechanicalId);
    }

    public BlokemonRuntimeManifest Manifest { get; }

    public bool Contains(MechanicalCardId id) =>
        _collectibles.ContainsKey(id.Value)
        || _kits.ContainsKey(id.Value)
        || _vim.ContainsKey(id.Value);

    public CardKind Kind(MechanicalCardId id) =>
        _collectibles.ContainsKey(id.Value) ? CardKind.Bloke
        : _kits.ContainsKey(id.Value) ? CardKind.Kit
        : CardKind.Vim;

    public int CopyLimit(MechanicalCardId id) =>
        _collectibles.TryGetValue(id.Value, out var bloke) ? bloke.StackCopyLimit
        : _kits.TryGetValue(id.Value, out var kit) ? kit.StackCopyLimit
        : _vim[id.Value].StackCopyLimit;

    public bool IsRegular(MechanicalCardId id) =>
        _collectibles.TryGetValue(id.Value, out var card) && card.Rank == BlokemonRank.Regular;

    public bool IsFossil(MechanicalCardId id) =>
        Manifest.BaseRules.FossilKits.KitIds.Contains(id.Value, StringComparer.Ordinal);

    public BlokemonCollectible Bloke(MechanicalCardId id) => _collectibles[id.Value];

    public BlokemonKit Kit(MechanicalCardId id) => _kits[id.Value];

    public BlokemonBasicVim Vim(MechanicalCardId id) => _vim[id.Value];

    public BlokemonAttack? Attack(EffectId id) => _attacks.GetValueOrDefault(id.Value);

    public BlokemonPartyTrick? PartyTrick(EffectId id) => _partyTricks.GetValueOrDefault(id.Value);

    public BlokemonHouseRule? HouseRule(EffectId id) => _houseRules.GetValueOrDefault(id.Value);

    public int StayingPower(CardState card) =>
        card.Kind == CardKind.Bloke
            ? Bloke(card.MechanicalId).StayingPower
            : Manifest.BaseRules.FossilKits.PlayAsRegularLocalStayingPower;

    public int TaxiFare(CardState card) =>
        card.Kind == CardKind.Bloke ? Bloke(card.MechanicalId).TaxiFare : int.MaxValue;

    public int BarChits(CardState card) =>
        card.Kind == CardKind.Bloke ? Bloke(card.MechanicalId).BarChitsWhenSentHome
        : Manifest.BaseRules.FossilKits.SentHomeAwardsOneBarChit ? 1
        : 0;

    public FrozenList<BlokemonMechanicalType> MechanicalTypes(CardState card) =>
        card.Kind == CardKind.Bloke
            ? FrozenList<BlokemonMechanicalType>.Create(Bloke(card.MechanicalId).MechanicalTypes)
            : FrozenList<BlokemonMechanicalType>.Create(BlokemonMechanicalType.Colorless);

    public IEnumerable<BlokemonPartyTrick> PartyTricks(CardState card) =>
        card.Kind == CardKind.Bloke ? Bloke(card.MechanicalId).PartyTricks
        : card.Kind == CardKind.Kit ? Kit(card.MechanicalId).PartyTricks
        : [];

    public IEnumerable<BlokemonAttack> Attacks(CardState card) =>
        card.Kind == CardKind.Bloke ? Bloke(card.MechanicalId).Attacks
        : card.Kind == CardKind.Kit ? Kit(card.MechanicalId).Attacks
        : [];

    public IEnumerable<BlokemonHouseRule> HouseRules(CardState card) =>
        card.Kind == CardKind.Bloke ? Bloke(card.MechanicalId).HouseRules
        : card.Kind == CardKind.Kit ? Kit(card.MechanicalId).HouseRules
        : [];
}

File: src/Blokemon.Game/AuthorityCatalog.fs
namespace Blokemon.Game

open System
open System.Collections.Generic
open System.Linq
open Blokemon.Core.SetDesign

/// Every lookup the engine and the interpreter make against the printed authority, indexed once
/// per engine instead of scanned per question.
type internal AuthorityCatalog(manifest: BlokemonRuntimeManifest) =
    // Add, not the indexer: a duplicated mechanical id is an authority defect and must surface as
    // one, exactly as the C# ToDictionary calls this replaces did.
    let index (keyOf: 'T -> string) (items: 'T seq) =
        let result = Dictionary<string, 'T>(StringComparer.Ordinal)

        for item in items do
            result.Add(keyOf item, item)

        result

    let collectibles = manifest.Collectibles |> index (fun card -> card.Id)
    let kits = manifest.Kits |> index (fun card -> card.Id)
    let vim = manifest.BasicVim |> index (fun card -> card.Id)

    let attacks =
        Seq.append
            (manifest.Collectibles |> Seq.collect (fun card -> card.Attacks))
            (manifest.Kits |> Seq.collect (fun card -> card.Attacks))
        |> index (fun effect -> effect.MechanicalId)

    let partyTricks =
        Seq.append
            (manifest.Collectibles |> Seq.collect (fun card -> card.PartyTricks))
            (manifest.Kits |> Seq.collect (fun card -> card.PartyTricks))
        |> index (fun effect -> effect.MechanicalId)

    let houseRules =
        Seq.append
            (manifest.Collectibles |> Seq.collect (fun card -> card.HouseRules))
            (manifest.Kits |> Seq.collect (fun card -> card.HouseRules))
        |> index (fun effect -> effect.MechanicalId)

    let tryGet (source: Dictionary<string, 'T>) (key: string) =
        match source.TryGetValue key with
        | true, value -> ValueSome value
        | _ -> ValueNone

    member _.Manifest = manifest

    member _.Contains(id: MechanicalCardId) =
        collectibles.ContainsKey id.Value
        || kits.ContainsKey id.Value
        || vim.ContainsKey id.Value

    member _.Kind(id: MechanicalCardId) =
        if collectibles.ContainsKey id.Value then CardKind.Bloke
        elif kits.ContainsKey id.Value then CardKind.Kit
        else CardKind.Vim

    member _.CopyLimit(id: MechanicalCardId) =
        match tryGet collectibles id.Value with
        | ValueSome bloke -> bloke.StackCopyLimit
        | ValueNone ->
            match tryGet kits id.Value with
            | ValueSome kit -> kit.StackCopyLimit
            | ValueNone -> vim[id.Value].StackCopyLimit

    member _.IsRegular(id: MechanicalCardId) =
        match tryGet collectibles id.Value with
        | ValueSome card -> card.Rank = BlokemonRank.Regular
        | ValueNone -> false

    member _.IsFossil(id: MechanicalCardId) =
        manifest.BaseRules.FossilKits.KitIds.Contains(id.Value, StringComparer.Ordinal)

    member _.Bloke(id: MechanicalCardId) = collectibles[id.Value]

    member _.Kit(id: MechanicalCardId) = kits[id.Value]

    member _.Vim(id: MechanicalCardId) = vim[id.Value]

    member _.Attack(id: EffectId) = tryGet attacks id.Value

    member _.PartyTrick(id: EffectId) = tryGet partyTricks id.Value

    member _.HouseRule(id: EffectId) = tryGet houseRules id.Value

    member this.StayingPower(card: CardState) =
        if card.Kind = CardKind.Bloke then
            this.Bloke(card.MechanicalId).StayingPower
        else
            manifest.BaseRules.FossilKits.PlayAsRegularLocalStayingPower

    member this.TaxiFare(card: CardState) =
        if card.Kind = CardKind.Bloke then
            this.Bloke(card.MechanicalId).TaxiFare
        else
            Int32.MaxValue

    member this.BarChits(card: CardState) =
        if card.Kind = CardKind.Bloke then
            this.Bloke(card.MechanicalId).BarChitsWhenSentHome
        elif manifest.BaseRules.FossilKits.SentHomeAwardsOneBarChit then
            1
        else
            0

    member this.MechanicalTypes(card: CardState) =
        if card.Kind = CardKind.Bloke then
            FrozenList<BlokemonMechanicalType>.Create(this.Bloke(card.MechanicalId).MechanicalTypes)
        else
            FrozenList<BlokemonMechanicalType>.Create BlokemonMechanicalType.Colorless

    member this.PartyTricks(card: CardState) : BlokemonPartyTrick seq =
        if card.Kind = CardKind.Bloke then
            this.Bloke(card.MechanicalId).PartyTricks
        elif card.Kind = CardKind.Kit then
            this.Kit(card.MechanicalId).PartyTricks
        else
            Array.empty

    member this.Attacks(card: CardState) : BlokemonAttack seq =
        if card.Kind = CardKind.Bloke then
            this.Bloke(card.MechanicalId).Attacks
        elif card.Kind = CardKind.Kit then
            this.Kit(card.MechanicalId).Attacks
        else
            Array.empty

    member this.HouseRules(card: CardState) : BlokemonHouseRule seq =
        if card.Kind = CardKind.Bloke then
            this.Bloke(card.MechanicalId).HouseRules
        elif card.Kind = CardKind.Kit then
            this.Kit(card.MechanicalId).HouseRules
        else
            Array.empty
